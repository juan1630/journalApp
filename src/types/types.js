export const types = {
    loggin: '[auth] Login',
    logout: '[auth] Logout',

    uiRemoveError: '[UI] Remove Error',
    uiSetError: '[UI] Set Error'

}