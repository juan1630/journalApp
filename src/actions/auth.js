// imports de otras despencias 
import Swal from 'sweetalert2';

import { types } from "../types/types";
import { firebase, googleAuthProvider } from '../firebase/firebase-config'
import { startLoading, finishLoading } from './ui'
import { noteLogout } from './notes';


export const loginWithEmailYPassword = (email, password) => {
    
    // esta funcion retorna un callback
    return (dispatch) => {

        dispatch( startLoading() );

        return firebase.auth().signInWithEmailAndPassword(email, password)
        .then(({user}) => {

            dispatch( finishLoading() );
            dispatch( login( user.uid, user.displayName ) );
        })
        .catch( e => {
            Swal.fire('Fail', e.message, 'error');
            dispatch(finishLoading() )
        });
    }
}


export const startGoogleLgoin = () => {
    
    return (dispatch) => {

        //hacemos uso de la funcio auth y el provider es google que esta declarado en el config de firebase
        firebase.auth().signInWithPopup( googleAuthProvider ) 
        .then( ({user})  => {

            dispatch( login( user.uid, user.displayName ));
        })
        .catch( (err) => {
            Swal.fire('Fail', err.message, 'error');
        });
    }
}


export const starRegisterWithEmailPasswordEmail = (name, email, password) => {

    return (dispatch) => {
         firebase.auth().createUserWithEmailAndPassword(email, password)
         .then(  async ({user}) => {

            // actualizamos el displayName porque viene en nulo cuando se hace el registro
          await  user.updateProfile({displayName: name})
                dispatch( login( user.uid, user.displayName) );
         })
         .catch( error  => console.error(error))
    }
}




export const startLogout = () => {
    return async (dispatch) => {
        // esta funcion regresa una promesa
        await firebase.auth().signOut();
        
        dispatch(logOut());
        dispatch( noteLogout() );
    }
}



export const login = (uid, displayName) =>  ({
        type: types.loggin,
        payload: {
            uid,
            displayName,
        }
});

export const logOut = () =>({
    type: types.logout
})