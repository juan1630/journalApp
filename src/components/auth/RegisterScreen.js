import React from 'react';
import {  Link } from 'react-router-dom';

export const RegisterScreen = () => {
    return (
        <div>
            <h3 className="auth__title"> Register </h3>

            <form>

                <input type="text"  className="auth__input" placeholder="Name" name="name" autoComplete="off" />

                <input type="email"  className="auth__input" placeholder="Email" name="email" autoComplete="off" />

                <input type="password" className="auth__input" placeholder="Password" name="pass" />

                <input type="password" className="auth__input" placeholder="Password confirm" name="pass2" />


                <button
                    className="btn btn-primary btn-block mb-5"
                    type="submit"
                >
                  Register
                </button>

                <Link to="/auth/register" className="link">
                    Already register
                </Link>

        </form>
    </div>
    )
}
