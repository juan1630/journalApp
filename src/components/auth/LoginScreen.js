import React from 'react';
import { Link } from 'react-router-dom';

import { useForm } from '../../hooks/useForm';
import { useDispatch } from 'react-redux';
import { login, loginWithEmailYPassword, startGoogleLgoin } from '../../actions/auth';


export const LoginScreen = () => {

    // usamos el dispatch de react redux
     const dispatch = useDispatch();

    // hace uso del hook useform
    const [ formValues, handleInputChange ] = useForm({

        email: 'Juan@gmail.com',
        password: '123456'
    });


    const { email, password } = formValues;

    const handleLoginGoogle = () => {
        dispatch( startGoogleLgoin() );
    }

    const handleLogin = (e) => {
        e.preventDefault();
        dispatch( loginWithEmailYPassword( email, password ) );
    }

    return (
        <div>
            <h3 className="auth__title"> Login Screen </h3>

            <form onSubmit={handleLogin} >

                <input type="text"  className="auth__input" 
                        placeholder="Email" 
                        name="email" autoComplete="off"
                        value= { email}
                        />

                <input type="password" className="auth__input" 
                        placeholder="Password" 
                        name="pass"
                        value={ password }
                        />
                
                <button
                    className="btn btn-primary btn-block"
                    type="submit"
                >
                    Login
                </button>

                <div  className="auth__social-networks" >
                    <p>Login with social networks</p>
                    <div 
                        className="google-btn"
                        onClick={ handleLoginGoogle }
                    >
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>

                <Link to="/auth/register" className="link" >
                    Create a new account
                </Link>
           
            </form>
        </div>
    )
}
