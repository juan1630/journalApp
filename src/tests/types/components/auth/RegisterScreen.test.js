

import React from 'react';

import { mount } from "enzyme";

import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { RegisterScreen } from "../../../../components/auth/RegisterScreen";


import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk'; 
import '@testing-library/jest-dom';
import { types } from '../../../../types/types';

// jest.mock('../../../../actions/auth', () => ({

//     startGoogleLgoin: jest.fn(),
//     loginWithEmailYPassword: jest.fn()
// }));
// creamos un mock de la funcion que vamos a llamar


const middlewares = [ thunk ];

// mocks 
const mockStore = configureStore(middlewares);

const initState = {
    auth: {},
    ui: {
        loading: false,
        msgError: null
    }
};


let store = mockStore(initState);
// store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={store} >
        <MemoryRouter>
            <RegisterScreen />
        </MemoryRouter>
    </Provider>
);



describe('Pruebas en el registerScreen', () => {
    

    test('Debe de hacer match con el snapshot', () => {

        expect( wrapper ).toMatchSnapshot();
    });
    


    test('Debe de hacer el dispatch de la accion respectiva', () => {

        const emailField = wrapper.find('input[name="email"]');
        // console.log( emailField );

        emailField.simulate('change', {
            target: {
                value: '',
                name: 'email'
            }

        })

        wrapper.find('form').simulate('submit', { 
            preventDefault(){}
        });


        const actions = store.getActions();
        // console.log(actions);

        expect( actions[0]).toEqual({
            type: types.uiSetError,
            payload: 'Email is not validate'
        });

    });


    test('Debe de mostrar la caja de alerta con el Error', () => {

        const initState = {
            auth: {},
            ui: {
                loading: false,
                msgError: "El email no es correcto"
            }
        };
        
        
        const store = mockStore(initState);
        // store.dispatch = jest.fn();
        
        const wrapper = mount(
            <Provider store={store} >
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        );
        
            expect( wrapper.find('.auth__alert-error').exists() ).toBe( true );
            expect( wrapper.find('.auth__alert-error').text() ).toBe( initState.ui.msgError );

    });

});