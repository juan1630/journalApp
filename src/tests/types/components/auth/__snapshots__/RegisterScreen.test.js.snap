// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el registerScreen Debe de hacer match con el snapshot 1`] = `
<div>
  <h3
    className="auth__title"
  >
     Register 
  </h3>
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="nameUser"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="Juan"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="josejuanpatron1630@gmail.com"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="123456"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Password confirm"
      type="password"
      value="123456"
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Register
    </button>
    Already register
    <a
      className="link"
      href="/auth/register"
      onClick={[Function]}
    />
  </form>
</div>
`;
